### DISCUSSION TITLE
Implementation of macOS window-decorations configuration option, and potential for extra customizability

### DISCUSSION THREAD

**Discussion by johnseth97** [ghostty-org/ghostty#2465] [Feature Requests, Ideas]
[Related PR](https://github.com/ghostty-org/ghostty/pull/2459) (currently closed to move to discussion format per @mitchellh)

Most terminal emulators on MacOS that support the removal of window decorations do such by hiding the navigation controls and title, while making the titlebar transparent. 

It's an aesthetic choice, but I believe that it does keep Ghostty's style more in line with the rounded look found on macOS. And it does behave differently than the `macos-titlebar-style = hidden` option, [as addressed in this thread](https://github.com/ghostty-org/ghostty/discussions/2465#discussioncomment-10993875).

<img width="1512" alt="image" src="https://github.com/user-attachments/assets/ff15389a-d110-4806-a634-e8069a835eca">

I do understand that some people don't like the extra padding caused by hiding the bar instead of removing it altogether, so another option would be to add multiple options for window-decorations on macOS.

`window-decorations = none`:
```swift
if (!ghostty.config.windowDecorations) { window.styleMask.remove(.titled) }
```
`window-decorations = none_rounded`:
```swift
if (!ghostty.config.windowDecorations) {
    window.titleVisibility = .hidden
    window.standardWindowButton(.closeButton)?.isHidden = true
    window.standardWindowButton(.miniaturizeButton)?.isHidden = true
    window.standardWindowButton(.zoomButton)?.isHidden = true
    window.titlebarAppearsTransparent = true
    window.styleMask.insert(.fullSizeContentView)
}
```
`window-decorations = transparent`:
```swift
if (!ghostty.config.windowDecorations) {
    window.titlebarAppearsTransparent = true
}
```

etc.

Additional commentary from @mitchellh in PR: 

> I have absolutely nothing against the technical implementation of this PR, but I do wonder what the correct UX in terms of configuration should be to expose this. I think turning window-decoration into an enum with some extra options may make sense.

**Comment by qwerasd205** [comment]
Did you read [the docs](https://github.com/ghostty-org/ghostty/blob/main/src/config/Config.zig#L894-L908) for `window-decorations`, which point to [`macos-titlebar-style = hidden`](https://github.com/ghostty-org/ghostty/blob/main/src/config/Config.zig#L1415-L1445) which covers this behavior already?

**Reply by johnseth97** [reply]
Admittedly, I was not aware of the hidden behavior. I apologize for not reading the documentation.

However, the resulting behavior is different than `macos-titlebar-style = hidden`, notably by keeping any resulting TUI elements out of the unclickable titlebar space, a problem described in the linked documentation.

Here's a screenshot of the same TUI setup as the parent discussion post, demonstrating the difference in behavior. With the `macos-titlebar-style = hidden` option, the tmux status bar butts up against the top of the window, as would the tab-bar for Neovim, blocking mouse input on both. 

<img width="1512" alt="image" src="https://github.com/user-attachments/assets/2f8a4e48-1e13-41be-9f88-258062541e75">

Additionally, with the proposed behavior(which should probably be implemented as a `macos-titlebar-style` option instead), there is the potential ability to manually change the height of the titlebar in config, which would change the size of the padding on top. 

I still can't wrap my head around exactly why they behave different in the code, as I haven't spent enough time looking at the `TerminalWindow` swift file.

TLDR: PR is poorly implemented and I should read the docs better, but the resulting behavior is different and solves a significant TUI related problem present in the existing `macos-titlebar-style = hidden` config option.
